chainio/blockbeat.go

```diff
@@ -0,0 +1,154 @@
+// error chan.
+func NewBeat(epoch chainntnfs.BlockEpoch) Beat {
+	b := Beat{
+		epoch:   epoch,
+		errChan: make(chan error, 1),
+	}
+
+	// Create a customized logger for the blockbeat.
+	logPrefix := fmt.Sprintf("Height[%6d]:", b.Height())
+	b.log = build.NewPrefixLog(logPrefix, clog)
+
+	return b
+}
+
+// NewBeatFromHeight creates a new beat with the specified height and a
+// buffered error chan. This method should only be used in test.
+func NewBeatFromHeight(height int32) Beat {
+	return NewBeat(chainntnfs.BlockEpoch{Height: height})
+}
+
+// Height returns the height of the block epoch.
+//
+// NOTE: Part of the Blockbeat interface.
+func (b Beat) Height() int32 {
+	return b.epoch.Height
+}
+
+// NOTE: Part of the Blockbeat interface.
+func (b Beat) NotifyBlockProcessed(err error, quitChan chan struct{}) {
+	fn.SendOrQuit(b.errChan, err, quitChan)
+}
+
+// DispatchSequential takes a list of consumers and notify them about the new
+// epoch sequentially.
+//
+// NOTE: Part of the Blockbeat interface.
+func (b Beat) DispatchSequential(consumers []Consumer) error {
+	for _, c := range consumers {
+		// Send the copy of the beat to the consumer.
+		if err := b.notifyAndWait(c); err != nil {
+			b.log.Errorf("Consumer=%v failed to process "+
+				"block: %v", c.Name(), err)
+
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DispatchConcurrent notifies each consumer concurrently about the blockbeat.
+//
+// NOTE: Part of the Blockbeat interface.
+func (b Beat) DispatchConcurrent(consumers []Consumer) error {
+	// errChans is a map of channels that will be used to receive errors
+	// returned from notifying the consumers.
+	errChans := make(map[string]chan error, len(consumers))
+
+	// Notify each queue in goroutines.
+	for _, c := range consumers {
+		// Create a signal chan.
+		errChan := make(chan error, 1)
+		errChans[c.Name()] = errChan
+
+		// Notify each consumer concurrently.
+		go func(c Consumer, b Beat) {
+			// Send the copy of the beat to the consumer.
+			errChan <- b.notifyAndWait(c)
+		}(c, b)
+	}
+
+	// Wait for all consumers in each queue to finish.
+	for name, errChan := range errChans {
+		err := <-errChan
+		if err != nil {
+			b.log.Errorf("Consumer=%v failed to process block: %v",
+				name, err)
+
+			return err
+		}
+	}
+
+	return nil
+}
+
+// notifyAndWait sends the blockbeat to the specified consumer. It requires the
+// consumer to finish processing the block under 30s, otherwise a timeout error
+// is returned.
+func (b Beat) notifyAndWait(c Consumer) error {
+	// Construct a new beat with a buffered error chan.
+	beatCopy := NewBeat(b.epoch)
+
+	b.log.Debugf("Waiting for consumer[%v] to process it", c.Name())
+
+	// Record the time it takes the consumer to process this block.
+	start := time.Now()
+
+	// We expect the consumer to finish processing this block under 30s,
+	// otherwise a timeout error is returned.
+	err, timeout := fn.RecvOrTimeout(
+		c.ProcessBlock(beatCopy), DefaultProcessBlockTimeout,
+	)
```

Comment:
I _think_ we may need 2 separate timeout checks here: 
- one for actually getting back the error channel from `c.ProcessBlock(beatCopy)`
- another waiting for a signal on that returned error channel.

This currently does the second part but I think what can happen is that `c.ProcessBLock` hangs and so then `RecvOrTimeout` never actually starts?

Reply:
ie, I think we actually need a helper like this:

```
func RecvOrTimeout[T any](fn func() <-chan T, timeout time.Duration) (T, error) {
	var (
		c           chan T
		timeoutChan = time.After(timeout)
	)

	// Wait for the function to return the channel we want to listen on.
	select {
	case c = <-fn():
	case <-timeoutChan:
		var zero T
		return zero, fmt.Errorf("timeout hit")
	}

	// Now that we have the channel, wait to receive on it.
	select {
	case m := <-c:
		return m, nil

	case <-timeoutChan:
		var zero T
		return zero, fmt.Errorf("timeout hit")
	}
}
```